src/net_processing.cpp

```diff
@@ -2405,6 +2405,9 @@ void ProcessMessage(
                       pfrom.m_tx_relay == nullptr ? "block-relay" : "full-relay");
         }
 
+        // Signal ADDRv2 support (BIP155).
+        connman->PushMessage(&pfrom, msgMaker.Make(NetMsgType::SENDADDRv2));
```

Comment:
The first point is easy to fix:

<details>
<summary>patch to change the order of initial messages</summary>

```diff
diff --git i/src/net_processing.cpp w/src/net_processing.cpp
index 5af63acc4..296237075 100644
--- i/src/net_processing.cpp
+++ w/src/net_processing.cpp
@@ -2318,12 +2318,15 @@ void ProcessMessage(
         // Be shy and don't send version until we hear
         if (pfrom.fInbound)
             PushNodeVersion(pfrom, connman, GetAdjustedTime());
 
         connman.PushMessage(&pfrom, CNetMsgMaker(INIT_PROTO_VERSION).Make(NetMsgType::VERACK));
 
+        // Signal ADDRv2 support (BIP155).
+        connman.PushMessage(&pfrom, CNetMsgMaker(INIT_PROTO_VERSION).Make(NetMsgType::SENDADDRv2));
+
         pfrom.nServices = nServices;
         pfrom.SetAddrLocal(addrMe);
         {
             LOCK(pfrom.cs_SubVer);
             pfrom.cleanSubVer = cleanSubVer;
         }
@@ -2428,15 +2431,12 @@ void ProcessMessage(
             LogPrintf("New outbound peer connected: version: %d, blocks=%d, peer=%d%s (%s)\n",
                       pfrom.nVersion.load(), pfrom.nStartingHeight,
                       pfrom.GetId(), (fLogIPs ? strprintf(", peeraddr=%s", pfrom.addr.ToString()) : ""),
                       pfrom.m_tx_relay == nullptr ? "block-relay" : "full-relay");
         }
 
-        // Signal ADDRv2 support (BIP155).
-        connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::SENDADDRv2));
-
         if (pfrom.nVersion >= SENDHEADERS_VERSION) {
             // Tell our peer we prefer to receive headers rather than inv's
             // We send this to non-NODE NETWORK peers as well, because even
             // non-NODE NETWORK peers can announce blocks (such as pruning
             // nodes)
             connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::SENDHEADERS));
```
</details>

would change the order of sent messages from `verack, getaddr, sendaddrv2` to `verack, sendaddrv2, getaddr`.

For the second point - the current code in this PR will assume no-addrv2 support when `CNode::PushAddress()` is called early when advertising our address. As a result if our address is e.g. TORv3, then it will not be appended to `vAddrToSend[]`. Hmm...

Reply:
On the second point:

> we also add our own address to vAddrToSend on receipt of the version message to advertise our address to the peer.

Right, that is in `AdvertiseLocal()`, called from `PeerLogicValidation::SendMessages()`

> That gets sent out in an addr message in the SendMessages() loop on the m_next_addr_send loop.

Not necessary as an `addr` message - if we have received `sendaddrv2` and thus set `CNode::m_wants_addrv2` to `true`, then we will send `addrv2` message. But see below, you are right that on the first iteration we still have not received `sendaddrv2`.

> I expect that will be on the first call time through SendMessages() since m_next_addr_send is initialized to 0.

Right.

> That means we'll advertise our local address using an addr message because we won't have received our peer's sendaddrv2 message yet.

Yes, for the first call to `AdvertiseLocal()` and only if our address is `addr`-compatible. Notice the following:

* `AdvertiseLocal()` is being called periodically (`AVG_LOCAL_ADDRESS_BROADCAST_INTERVAL`)
* If our address is `addr`-compatible, then broadcasting it in `addr` message is fine, even if the peer has asked for `addrv2` messages via `sendaddrv2`.
* If our address is not `addr`-compatible (e.g. TORv3) then the first call to `AdvertiseLocal()` will end up being a noop because it will call `CNode::PushAddress()` which will see that the address being pushed is not supported by the peer and so it will not add it to `vAddrToSend[]`.
  * On the next call to `AdvertiseLocal()` we should already have received `sendaddrv2` and thus be aware that the peer supports `addrv2` and so will properly advertise our address in `addrv2` message.

**Verdict**
A "problem" exists only if our address is not `addrv2`-compatible (e.g. TORv3) and that "problem" manifests as the advertising of our address to the peer will not be done immediately when the connection is established, but some time later.

Is that acceptable?

Reply:
I haven't reviewed the implementation, but if it's as you say:

> - If our address is not addr-compatible (e.g. TORv3) then the first call to AdvertiseLocal() will end up being a noop because it will call CNode::PushAddress() which will see that the address being pushed is not supported by the peer and so it will not add it to vAddrToSend[].
>   - On the next call to AdvertiseLocal() we should already have received sendaddrv2 and thus be aware that the peer supports addrv2 and so will properly advertise our address in addrv2 message.

then I think that's fine.

(My note was more something to be aware of than saying that this would be a serious issue)

Reply:
Yes, here are the changes to `PushAddress()`:

https://github.com/bitcoin/bitcoin/pull/19031/files#diff-1a8b9d1ad0a6fda5e751286c73102fc2R953

It will only append the address to `vAddrToSend[]` if it is supported by the peer (which is - either the peer supports `addrv2` or he does not but the address is `addr`-compatible).

We can even avoid the delay if we detect that `AdvertiseLocal()` did nothing (due to `addrv2` address and peer that does not support `addrv2` (yet)) and if that is the case then don't update `m_next_local_addr_send` so that the next iteration will also call `AdvertiseLocal()`.

Reply:
I addressed the first concern and now the order of sent messages is `verack, sendaddrv2, getaddr`.